src/ethereum/prague/vm/__init__.py

```diff
@@ -0,0 +1,224 @@
+"""
+Ethereum Virtual Machine (EVM)
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+.. contents:: Table of Contents
+    :backlinks: none
+    :local:
+
+Introduction
+------------
+
+The abstract computer which runs the code stored in an
+`.fork_types.Account`.
+"""
+
+import enum
+from dataclasses import dataclass
+from typing import List, Optional, Set, Tuple, Union
+
+from ethereum.base_types import U64, U256, Bytes, Bytes0, Bytes32, Uint
+from ethereum.crypto.hash import Hash32
+
+from ..blocks import Log
+from ..fork_types import Address, VersionedHash
+from ..state import State, TransientStorage, account_exists_and_is_empty
+from .exceptions import InvalidEOF
+from .precompiled_contracts import RIPEMD160_ADDRESS
+
+__all__ = ("Environment", "Evm", "Message", "EOF")
+
+
+EOF_MAGIC = b"\xEF\x00"
+EOF_MAGIC_LENGTH = 2
+
+MAX_CODE_SIZE = 0x6000
+
+
+class EOF(enum.Enum):
+    """
+    Enumeration of the different kinds of EOF containers.
+    Legacy code is assigned zero.
+    """
+
+    LEGACY = 0
+    EOF1 = 1
+
+
+@dataclass
+class EOFMetadata:
+    """
+    Dataclass to hold the metadata information of the
+    EOF container.
+    """
+
+    type_size: Uint
+    num_code_sections: Uint
+    code_sizes: List[Uint]
+    num_container_sections: Uint
+    container_sizes: Optional[List[Uint]]
+    data_size: Uint
+    body_start_index: Uint
+    type_section_contents: List[bytes]
+    code_section_contents: List[bytes]
+    container_section_contents: Optional[List[bytes]]
+    data_section_contents: bytes
+
+
+@dataclass
+class ReturnStackItem:
+    """
+    Stack item for the return stack.
+    """
+
+    code_section_index: Uint
+    offset: Uint
+
+
+@dataclass
+class Environment:
+    """
+    Items external to the virtual machine itself, provided by the environment.
+    """
+
+    caller: Address
+    block_hashes: List[Hash32]
+    origin: Address
+    coinbase: Address
+    number: Uint
+    base_fee_per_gas: Uint
+    gas_limit: Uint
+    gas_price: Uint
+    time: U256
+    prev_randao: Bytes32
+    state: State
+    chain_id: U64
+    traces: List[dict]
+    excess_blob_gas: U64
+    blob_versioned_hashes: Tuple[VersionedHash, ...]
+    transient_storage: TransientStorage
+
+
+@dataclass
+class Message:
+    """
+    Items that are used by contract creation or message call.
+    """
+
+    caller: Address
+    target: Union[Bytes0, Address]
+    current_target: Address
+    gas: Uint
+    value: U256
+    data: Bytes
+    code_address: Optional[Address]
+    container: Bytes
+    depth: Uint
+    should_transfer_value: bool
+    is_static: bool
+    accessed_addresses: Set[Address]
+    accessed_storage_keys: Set[Tuple[Address, Bytes32]]
+    parent_evm: Optional["Evm"]
+
+
+@dataclass
+class Evm:
+    """The internal state of the virtual machine."""
+
+    pc: Uint
+    stack: List[U256]
+    memory: bytearray
+    code: Bytes
```

Comment:
Should we create [NewType](https://docs.python.org/3/library/typing.html#typing.NewType) for code vs. container vs. whatever other types of sections there are?

Reply:
I am interpreting this as the code being implemented. If it is legacy, then it's just the code. If EOF, it is the code in the section being currently implemented.